File: Algorithm/Python/insert_interval.py
class Intervals:
    def __init__(self,s=0,e=0):
        self.start=s
        self.end=e
class Solution:
    def insert(self,intervals,newInterval):
        n = len(intervals)
        if n == 0:
            return [newInterval]
        res=[]
        i = 0
        while i < n and  intervals[i].end < newInterval.start :
            res.append(intervals[i])
            i += 1
        start = newInterval.start
        end = newInterval.end
        while i <n and intervals[i].start <= newInterval.end :
            start = min(start, intervals[i].start)
            end = max(end, intervals[i].end)
            i += 1
        res.append(Intervals(start, end))
        while i < n:
            res.append(intervals[i])
            i += 1
        return res
if __name__=='__main__':
    sol=Solution()
    intervals=[]
    newInterval=[]
    intervals.append(Intervals(1,2))
    intervals.append(Intervals(3,5))
    intervals.append(Intervals(6,7))
    intervals.append(Intervals(8,10))
    intervals.append(Intervals(12,16))
    newInterval=Intervals(4,9) 
    #intervals.append(Intervals(1,5))
    #newInterval = Intervals(2,3)
    #intervals.append(Intervals(1,5))
    #newInterval = Intervals(6, 8)
      
          
    
    for fuck in sol.insert(intervals,newInterval):
        print [fuck.start,fuck.end]

File: Algorithm/CPP/insert_interval.cpp
# include<stdio.h>
# include <iostream>
# include <string>
# include <array>
# include <vector>
using namespace std;

struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};
class Solution {
public:

	vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) {
		
		if (intervals.empty()) return vector<Interval>{newInterval};
		vector<Interval> res;
		int i = 0;
		while (i < intervals.size() && intervals[i].end < newInterval.start)
		{
			res.push_back(intervals[i]);
			i++;
		}
		res.push_back(newInterval);
		if (i != intervals.size())
		{
			if (intervals[i].start < res.back().start) res.back().start = intervals[i].start;
		}
		
		for (; i < intervals.size(); ++i)
		{
			if (res.back().end < intervals[i].start)
			{
				res.push_back(intervals[i]);
			}
			else{
				res.back().end = max(res.back().end,intervals[i].end);
			}
		}
		return res;
		
	}
};


int main()

{
	vector<Interval> intervals = { Interval(1, 2), Interval(3, 5), Interval(6, 7), Interval(8, 10) ,Interval(12,16)};
	Solution sol;
	for (auto k : sol.insert(intervals,Interval(4,9)))
	{
		cout << "(" << k.start << ","<<k.end << ")" << endl;
	}
	system("pause");
}

